[MediaStream] Add support for torch

eric-carlson opened the pull request:
#### 9c0b105d80212675839a00325b58df55f408f1c0
<pre>
[MediaStream] Add support for torch
<a href="https://bugs.webkit.org/show_bug.cgi?id=262131">https://bugs.webkit.org/show_bug.cgi?id=262131</a>
rdar://116066837

Reviewed by Youenn Fablet.

Add support for `torch` to MediaTrackCapabilities and MediaTrackSettings, and expose
&apos;torch&apos; as a supported constraint in MockRealtimeVideoSource and AVVideoCaptureSource.

* LayoutTests/fast/mediastream/MediaStreamTrack-getCapabilities-expected.txt:
* LayoutTests/fast/mediastream/mediastreamtrack-video-torch-expected.txt: Added.
* LayoutTests/fast/mediastream/mediastreamtrack-video-torch.html: Added.
* Source/WebCore/Modules/mediastream/MediaDevices.cpp:
(WebCore::hasInvalidGetDisplayMediaConstraint):
* Source/WebCore/Modules/mediastream/MediaStreamTrack.cpp:
(WebCore::MediaStreamTrack::getSettings const):
* Source/WebCore/Modules/mediastream/MediaStreamTrack.h:
* Source/WebCore/Modules/mediastream/MediaStreamTrack.idl:
* Source/WebCore/Modules/mediastream/MediaTrackCapabilities.cpp:
(WebCore::toMediaTrackCapabilities):
* Source/WebCore/Modules/mediastream/MediaTrackCapabilities.h:
* Source/WebCore/Modules/mediastream/MediaTrackCapabilities.idl:
* Source/WebCore/Modules/mediastream/MediaTrackConstraints.cpp:
(WebCore::convertToInternalForm):
* Source/WebCore/Modules/mediastream/MediaTrackConstraints.h:
* Source/WebCore/Modules/mediastream/MediaTrackConstraints.idl:
* Source/WebCore/Modules/mediastream/MediaTrackSupportedConstraints.h:
* Source/WebCore/Modules/mediastream/MediaTrackSupportedConstraints.idl:
* Source/WebCore/PAL/pal/cocoa/AVFoundationSoftLink.h:
* Source/WebCore/PAL/pal/cocoa/AVFoundationSoftLink.mm:
* Source/WebCore/platform/mediastream/MediaConstraints.cpp:
(WebCore::MediaTrackConstraintSetMap::filter const):
(WebCore::MediaTrackConstraintSetMap::set):
* Source/WebCore/platform/mediastream/MediaConstraints.h:
(WebCore::MediaTrackConstraintSetMap::aspectRatio const):
(WebCore::MediaTrackConstraintSetMap::facingMode const):
(WebCore::MediaTrackConstraintSetMap::whiteBalanceMode const):
(WebCore::MediaTrackConstraintSetMap::zoom const):
(WebCore::MediaTrackConstraintSetMap::torch const):
* Source/WebCore/platform/mediastream/RealtimeMediaSource.cpp:
(WebCore::RealtimeMediaSource::fitnessDistance):
(WebCore::RealtimeMediaSource::applyConstraint):
(WebCore::RealtimeMediaSource::supportsConstraint):
(WebCore::RealtimeMediaSource::supportsConstraints):
(WebCore::RealtimeMediaSource::setTorch):
* Source/WebCore/platform/mediastream/RealtimeMediaSource.h:
* Source/WebCore/platform/mediastream/RealtimeMediaSourceCapabilities.h:
(WebCore::RealtimeMediaSourceCapabilities::RealtimeMediaSourceCapabilities):
(WebCore::RealtimeMediaSourceCapabilities::supportsWhiteBalanceMode const):
(WebCore::RealtimeMediaSourceCapabilities::whiteBalanceModes const):
(WebCore::RealtimeMediaSourceCapabilities::setWhiteBalanceModes):
(WebCore::RealtimeMediaSourceCapabilities::supportsTorch const):
(WebCore::RealtimeMediaSourceCapabilities::torch const):
(WebCore::RealtimeMediaSourceCapabilities::setTorch):
* Source/WebCore/platform/mediastream/RealtimeMediaSourceCenter.cpp:
(WebCore::RealtimeMediaSourceCenter::RealtimeMediaSourceCenter):
* Source/WebCore/platform/mediastream/RealtimeMediaSourceSettings.cpp:
(WebCore::RealtimeMediaSourceSettings::convertFlagsToString):
(WebCore::RealtimeMediaSourceSettings::difference const):
* Source/WebCore/platform/mediastream/RealtimeMediaSourceSettings.h:
(WebCore::RealtimeMediaSourceSettings::allFlags):
(WebCore::RealtimeMediaSourceSettings::RealtimeMediaSourceSettings):
(WebCore::RealtimeMediaSourceSettings::supportsWhiteBalanceMode const):
(WebCore::RealtimeMediaSourceSettings::whiteBalanceMode const):
(WebCore::RealtimeMediaSourceSettings::setWhiteBalanceMode):
(WebCore::RealtimeMediaSourceSettings::supportsTorch const):
(WebCore::RealtimeMediaSourceSettings::torch const):
(WebCore::RealtimeMediaSourceSettings::setTorch):
* Source/WebCore/platform/mediastream/RealtimeMediaSourceSupportedConstraints.cpp:
(WebCore::RealtimeMediaSourceSupportedConstraints::supportsConstraint const):
* Source/WebCore/platform/mediastream/RealtimeMediaSourceSupportedConstraints.h:
(WebCore::RealtimeMediaSourceSupportedConstraints::RealtimeMediaSourceSupportedConstraints):
(WebCore::RealtimeMediaSourceSupportedConstraints::supportsWhiteBalanceMode const):
(WebCore::RealtimeMediaSourceSupportedConstraints::setSupportsWhiteBalanceMode):
(WebCore::RealtimeMediaSourceSupportedConstraints::supportsTorch const):
(WebCore::RealtimeMediaSourceSupportedConstraints::setSupportsTorch):
* Source/WebCore/platform/mediastream/mac/AVVideoCaptureSource.h:
* Source/WebCore/platform/mediastream/mac/AVVideoCaptureSource.mm:
(WebCore::AVVideoCaptureSource::AVVideoCaptureSource):
(WebCore::AVVideoCaptureSource::settingsDidChange):
(WebCore::AVVideoCaptureSource::settings):
(WebCore::AVVideoCaptureSource::capabilities):
(WebCore::AVVideoCaptureSource::setSessionSizeFrameRateAndZoom):
(WebCore::AVVideoCaptureSource::lockForConfiguration):
(WebCore::AVVideoCaptureSource::updateWhiteBalanceMode):
(WebCore::AVVideoCaptureSource::updateTorch):
* Source/WebCore/platform/mock/MockMediaDevice.h:
* Source/WebCore/platform/mock/MockRealtimeMediaSourceCenter.cpp:
(WebCore::defaultDevices):
* Source/WebCore/platform/mock/MockRealtimeVideoSource.cpp:
(WebCore::MockRealtimeVideoSource::capabilities):
(WebCore::MockRealtimeVideoSource::settings):
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:

Canonical link: <a href="https://commits.webkit.org/268853@main">https://commits.webkit.org/268853@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2aecfba9596abaa2f2570b1602d64bcf634f024b

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/20869 "2 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/21277 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/21937 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/22759 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/19446 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/21106 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/24515 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/21453 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/20739 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/21091 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/20858 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/18130 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/23613 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/18022 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/18935 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/25230 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/19115 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/19130 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/23141 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/19698 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/16729 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/18940 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/18771 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/5005 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/23269 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/19513 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

youennf commented on Source/WebCore/Modules/mediastream/MediaStreamTrack.idl:
```diff
@@ -73,7 +73,27 @@ enum MediaStreamTrackState { "live", "ended" };
+
+    // FIXME: add focusDistance;
+    // FIXME: add pan;
+    // FIXME: add tilt;
```
Not sure we want pan and tilt

eric-carlson replied:
True, will remove

youennf commented on Source/WebCore/platform/mediastream/RealtimeMediaSource.cpp:
```diff
@@ -681,6 +659,37 @@ double RealtimeMediaSource::fitnessDistance(const MediaConstraint& constraint)
+        if (!capabilities.supportsTorch())
+            return 0;
+
+        return downcast<BooleanConstraint>(constraint).fitnessDistance(capabilities.torch());
```
Given we are uncertain about required constraints for extensions such as torch et al, maybe we should just ignore them for now with regards to fitness distance computation, and add a FIXME?

eric-carlson replied:
I will return 0 if it is a mandatory constraint and we can talk about whether or not this is the right strategy as we fill in the other attributes.

youennf commented on Source/WebCore/platform/mediastream/RealtimeMediaSourceCapabilities.h:
```diff
@@ -233,10 +230,18 @@ class RealtimeMediaSourceCapabilities {
 
+    bool supportsTorch() const { return m_supportedConstraints.supportsTorch(); }
+    bool torch() const { return m_torch; }
+    void setTorch(const bool torch) { m_torch = torch; }
```
s/const bool/bool/

eric-carlson replied:
Oops!

youennf commented on Source/WebCore/platform/mediastream/mac/AVVideoCaptureSource.mm:
```diff
@@ -617,27 +629,62 @@ static inline IntDegrees sensorOrientation(AVCaptureVideoOrientation videoOrient
 }
 ALLOW_DEPRECATED_DECLARATIONS_END
 
+bool AVVideoCaptureSource::lockForConfiguration()
```
Should we also use it inside AVVideoCaptureSource::setSessionSizeFrameRateAndZoom()?

eric-carlson replied:
I didn't initially because that is a bigger change, since we don't bail if we aren't able to lock, but I think configuration will fail in that case anyway so I might as well fix it now.

youennf commented on Source/WebCore/platform/mediastream/mac/AVVideoCaptureSource.mm:
```diff
@@ -617,27 +629,62 @@ static inline IntDegrees sensorOrientation(AVCaptureVideoOrientation videoOrient
+    if (!lockForConfiguration())
+        return;
+
+    NSError *error = nil;
```
Could be defined closer to where it is used.

eric-carlson replied:
Sure

youennf commented on Source/WebCore/platform/mock/MockRealtimeMediaSourceCenter.cpp:
```diff
@@ -76,7 +76,8 @@ static inline Vector<MockMediaDevice> defaultDevices()
                 Color::black,
-                { },
+                { }, // whiteBalanceModes
+                false, // supportsTorch
```
It would be good to add a test that uses that camera and check torch.

eric-carlson replied:
The new test does check this. `capabilities.torch` returns `undefined`, but that is the same pattern we currently use for unsupported capabilities on all capture devices. We should fix this, but I will leave it as is and we can fix them all at the same time.

youennf reviewed (approved):
LGTM once bots are happy. I would also add a test with capabilities.torch=false.

eric-carlson committed:
[MediaStream] Add support for torch
https://bugs.webkit.org/show_bug.cgi?id=262131
rdar://116066837

Reviewed by Youenn Fablet.

Add support for `torch` to MediaTrackCapabilities and MediaTrackSettings, and expose
'torch' as a supported constraint in MockRealtimeVideoSource and AVVideoCaptureSource.

* LayoutTests/fast/mediastream/MediaStreamTrack-getCapabilities-expected.txt:
* LayoutTests/fast/mediastream/mediastreamtrack-video-torch-expected.txt: Added.
* LayoutTests/fast/mediastream/mediastreamtrack-video-torch.html: Added.
* Source/WebCore/Modules/mediastream/MediaDevices.cpp:
(WebCore::hasInvalidGetDisplayMediaConstraint):
* Source/WebCore/Modules/mediastream/MediaStreamTrack.cpp:
(WebCore::MediaStreamTrack::getSettings const):
* Source/WebCore/Modules/mediastream/MediaStreamTrack.h:
* Source/WebCore/Modules/mediastream/MediaStreamTrack.idl:
* Source/WebCore/Modules/mediastream/MediaTrackCapabilities.cpp:
(WebCore::toMediaTrackCapabilities):
* Source/WebCore/Modules/mediastream/MediaTrackCapabilities.h:
* Source/WebCore/Modules/mediastream/MediaTrackCapabilities.idl:
* Source/WebCore/Modules/mediastream/MediaTrackConstraints.cpp:
(WebCore::convertToInternalForm):
* Source/WebCore/Modules/mediastream/MediaTrackConstraints.h:
* Source/WebCore/Modules/mediastream/MediaTrackConstraints.idl:
* Source/WebCore/Modules/mediastream/MediaTrackSupportedConstraints.h:
* Source/WebCore/Modules/mediastream/MediaTrackSupportedConstraints.idl:
* Source/WebCore/PAL/pal/cocoa/AVFoundationSoftLink.h:
* Source/WebCore/PAL/pal/cocoa/AVFoundationSoftLink.mm:
* Source/WebCore/platform/mediastream/MediaConstraints.cpp:
(WebCore::MediaTrackConstraintSetMap::filter const):
(WebCore::MediaTrackConstraintSetMap::set):
* Source/WebCore/platform/mediastream/MediaConstraints.h:
(WebCore::MediaTrackConstraintSetMap::aspectRatio const):
(WebCore::MediaTrackConstraintSetMap::facingMode const):
(WebCore::MediaTrackConstraintSetMap::whiteBalanceMode const):
(WebCore::MediaTrackConstraintSetMap::zoom const):
(WebCore::MediaTrackConstraintSetMap::torch const):
* Source/WebCore/platform/mediastream/RealtimeMediaSource.cpp:
(WebCore::RealtimeMediaSource::fitnessDistance):
(WebCore::RealtimeMediaSource::applyConstraint):
(WebCore::RealtimeMediaSource::supportsConstraint):
(WebCore::RealtimeMediaSource::supportsConstraints):
(WebCore::RealtimeMediaSource::setTorch):
* Source/WebCore/platform/mediastream/RealtimeMediaSource.h:
* Source/WebCore/platform/mediastream/RealtimeMediaSourceCapabilities.h:
(WebCore::RealtimeMediaSourceCapabilities::RealtimeMediaSourceCapabilities):
(WebCore::RealtimeMediaSourceCapabilities::supportsWhiteBalanceMode const):
(WebCore::RealtimeMediaSourceCapabilities::whiteBalanceModes const):
(WebCore::RealtimeMediaSourceCapabilities::setWhiteBalanceModes):
(WebCore::RealtimeMediaSourceCapabilities::supportsTorch const):
(WebCore::RealtimeMediaSourceCapabilities::torch const):
(WebCore::RealtimeMediaSourceCapabilities::setTorch):
* Source/WebCore/platform/mediastream/RealtimeMediaSourceCenter.cpp:
(WebCore::RealtimeMediaSourceCenter::RealtimeMediaSourceCenter):
* Source/WebCore/platform/mediastream/RealtimeMediaSourceSettings.cpp:
(WebCore::RealtimeMediaSourceSettings::convertFlagsToString):
(WebCore::RealtimeMediaSourceSettings::difference const):
* Source/WebCore/platform/mediastream/RealtimeMediaSourceSettings.h:
(WebCore::RealtimeMediaSourceSettings::allFlags):
(WebCore::RealtimeMediaSourceSettings::RealtimeMediaSourceSettings):
(WebCore::RealtimeMediaSourceSettings::supportsWhiteBalanceMode const):
(WebCore::RealtimeMediaSourceSettings::whiteBalanceMode const):
(WebCore::RealtimeMediaSourceSettings::setWhiteBalanceMode):
(WebCore::RealtimeMediaSourceSettings::supportsTorch const):
(WebCore::RealtimeMediaSourceSettings::torch const):
(WebCore::RealtimeMediaSourceSettings::setTorch):
* Source/WebCore/platform/mediastream/RealtimeMediaSourceSupportedConstraints.cpp:
(WebCore::RealtimeMediaSourceSupportedConstraints::supportsConstraint const):
* Source/WebCore/platform/mediastream/RealtimeMediaSourceSupportedConstraints.h:
(WebCore::RealtimeMediaSourceSupportedConstraints::RealtimeMediaSourceSupportedConstraints):
(WebCore::RealtimeMediaSourceSupportedConstraints::supportsWhiteBalanceMode const):
(WebCore::RealtimeMediaSourceSupportedConstraints::setSupportsWhiteBalanceMode):
(WebCore::RealtimeMediaSourceSupportedConstraints::supportsTorch const):
(WebCore::RealtimeMediaSourceSupportedConstraints::setSupportsTorch):
* Source/WebCore/platform/mediastream/mac/AVVideoCaptureSource.h:
* Source/WebCore/platform/mediastream/mac/AVVideoCaptureSource.mm:
(WebCore::AVVideoCaptureSource::AVVideoCaptureSource):
(WebCore::AVVideoCaptureSource::settingsDidChange):
(WebCore::AVVideoCaptureSource::settings):
(WebCore::AVVideoCaptureSource::capabilities):
(WebCore::AVVideoCaptureSource::setSessionSizeFrameRateAndZoom):
(WebCore::AVVideoCaptureSource::lockForConfiguration):
(WebCore::AVVideoCaptureSource::updateWhiteBalanceMode):
(WebCore::AVVideoCaptureSource::updateTorch):
* Source/WebCore/platform/mock/MockMediaDevice.h:
* Source/WebCore/platform/mock/MockRealtimeMediaSourceCenter.cpp:
(WebCore::defaultDevices):
* Source/WebCore/platform/mock/MockRealtimeVideoSource.cpp:
(WebCore::MockRealtimeVideoSource::capabilities):
(WebCore::MockRealtimeVideoSource::settings):
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:

Canonical link: https://commits.webkit.org/268853@main
